### DISCUSSION TITLE
Implementing a custom factor in steady-state calculation

### DISCUSSION THREAD

**Discussion by mistloin** [idaholab/moose#17999] [Q&A Modules: General] [answered]
Hi. I studied about "time-stepping" framework in moose like the below equation; (increase or cutoff timestep along with residual value).

 dt_new = dt_old * "Factor"      --> "Factor" adjusted along with residual value

I currently use the moose library to solve a steady-state equation, and I want to implement a similar concept as;

  X_new = X_old - "Factor" * Inv(J) * F     --> "Factor" adjusted along with residual vlaue; 
  
  where X is solution vector, J is a jacobian matrix, F is residual

I have searched the moose framework which supports this kind of concept, but I have found nothing yet.

Should I develop a custom user object or other functional files for my question?

Or is there any framework that supports the above concept? 

Also, is there any way to get residual value (printed on-screen after each iteration) in my kernel?
(I tried to approach the printed value, the value may be printed through "PetscOutput.C", but I failed to access them)

Thanks for reading my question. Have a good day.

**Comment by GiudGiud** [comment, marked as answer]
Hello

There is a timestepper that can grow the time step by a constant factor. See https://mooseframework.inl.gov/source/timesteppers/IterationAdaptiveDT.html

For the update, I m not sure why you would want to add a factor there. Could you please give a reference for this method?

In the kernel? The kernel is executed at every quadrature point, so you want the residual at every quadrature point (which the kernel is contributing to)? Or do you want the residual for the whole domain?

Best,
Guillaume

**Reply by mistloin** [reply]
Thanks for your response. 

The reason why I want to use the factor in a steady-state is that the factor may improve convergence of my problem.
When I solve Newton method using MATLAB, I found that the problem (which was not converged) was converged by using the factor. (Initially the factor is very small value such as 1e-10, and along with residual, the factor increases up to 1)

I want to access the residual value for whole domain, because I should decides the factor value along with whole domain residual.

**Reply by GiudGiud** [reply]
Ok that makes sense. Somehow I thought the factor would be bigger than 1. 
You can use relaxed Picard iterations to achieve the same effect. In those you will have:
X_new_picard = a * X_new + (1-a) * X_old = (a + 1 -a) * X_old  - a * invJ F

We currently do not support varying relaxation factors, though that feature would not be very hard all to add in executioners/PicardSolve.C. I would make the relaxation factor a function then. This function could be the same as the one used by the timestepper to control the timestep.

For getting the residual in a kernel, you can use a Residual postprocessor, then create a new kernel that can take it as an input, then use getPostprocessorValueByName to get the value. 
Or you can look at postprocessors/Residual.C and import some of that code into your kernel

Best,
Guillaume

**Reply by mistloin** [reply]
Thanks for your advice! I will study about the Picard iteration and the residual post-proccessor just right now.
